mvcc: use point tombstones when rolling back imports

jeffswenson opened the pull request:
Previously, import rollback used range tombstones to delete contiguous
ranges of keys in non-empty tables. This is a relatively niche
optimization. It also allows users to create point keys on top of the
range tombstones by writing to the table after the rollback completes.
Import rollback is the only source of point tombstones over range keys
in CRDB.

Now, import rollbacks in non-empty tables will use point keys.

Context in
https://docs.google.com/document/d/1g7AUN3jUbzpLw2yR9C6k3tSib6XUqxYiNFp6OrCTJ4U.

Release Note: none
Fixes: https://github.com/cockroachdb/cockroach/issues/149697

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/137924)
<!-- Reviewable:end -->

jeffswenson committed:
storage: create copy of the predicate mvcc tests

This creates a copy of the predicate mvcc tests so that we can
be sure we are not changing the behavior of the range tombstone based
predicate delete.

Release note: none
Informs: #149697

jeffswenson committed:
mvcc: use point tombstones when rolling back imports

Previously, import rollback used range tombstones to delete contiguous
ranges of keys in non-empty tables. This is a relatively niche
optimization. It also allows users to create point keys on top of the
range tombstones by writing to the table after the rollback completes.
Import rollback is the only source of point tombstones over range keys
in CRDB.

Now, import rollbacks in non-empty tables will use point keys. A
separate change is needed for rollbacks of empty tables to avoid range
keys.

Context in
https://docs.google.com/document/d/1g7AUN3jUbzpLw2yR9C6k3tSib6XUqxYiNFp6OrCTJ4U.

Release Note: none
Fixes: #149697

msbutler commented on pkg/storage/mvcc.go:
```diff
@@ -3705,6 +3706,216 @@ func MVCCDeleteRange(
+		// NOTE: We don't check for a range key covering the point key because the
+		// iterator has range key masking enabled.
+		if !hasPointKey {
+			iter.Next()
```
why not call `NextKey`?

msbutler reviewed (commented):
awesome work! mostly nits

msbutler reviewed (approved):
LGTM! somebody from storage should probably give this a look too

stevendanna commented on pkg/storage/mvcc.go:
```diff
@@ -3705,6 +3706,216 @@ func MVCCDeleteRange(
+		return nil
+	}
+
+	iter, err := newMVCCIterator(ctx, rw, endTime, true, true, IterOptions{
```
MVCCPredicateDeleteRanges uses an MVCCIncrementalIterator. Might be worth a comment on why we aren't using that here.

stevendanna reviewed (commented):
Nice, just read this through to stay up to date. Don't consider my review blocking.

- I noted we have BenchmarkMVCCDeleteRangeWithPredicate, do we want to make another version of this for the new function as well?  I have no idea how useful that benchmark is.

jbowens reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> @jbowens reviewed 4 of 13 files at r2, 2 of 13 files at r4, 12 of 13 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137924#-:-OfzqfLi3P8OAN9uFYzN:b-uwkowd)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @annrpom, @jeffswenson, @msbutler, @mw5h, and @stevendanna)



<!-- Sent from Reviewable.io -->
